File: python/src/jevcal/schema.py
from __future__ import annotations

from dataclasses import dataclass, field as dataclass_field

from pydantic import BaseModel, Field, create_model

from .types import ConfidenceStrategy


def confidence_field_key(field_name: str) -> str:
    return f"{field_name}_confidence"


@dataclass(frozen=True)
class DecisionSchema:
    name: str
    version: str
    confidence_strategy: ConfidenceStrategy
    # Possibly augmented with `${field}_confidence` companion fields — usable
    # anywhere a plain Pydantic model is expected; safe to ignore the extra
    # keys outside jevcal, since the router strips them back out.
    wire_model: type[BaseModel]
    # The original, un-augmented field model — always safe to validate a
    # backend's cleaned value dict against.
    values_model: type[BaseModel]
    field_names: list[str] = dataclass_field(default_factory=list)


def define_decision(
    *,
    name: str,
    fields: type[BaseModel],
    version: str = "1",
    confidence: ConfidenceStrategy = "self-reported",
) -> DecisionSchema:
    """Defines a typed decision. `fields` is a plain Pydantic model — usable
    directly anywhere a schema is expected, no jevcal involvement required.

    `confidence` defaults to "self-reported" since it's the only strategy
    portable to every backend. When set, `wire_model` transparently gains a
    `${field}_confidence` companion field per top-level key.
    """
    field_names = list(fields.model_fields.keys())

    wire_model: type[BaseModel] = fields
    if confidence == "self-reported":
        extra_fields = {
            confidence_field_key(field_name): (
                float,
                Field(
                    ge=0,
                    le=1,
                    description=f'Self-rated confidence (0-1) that the "{field_name}" field above is correct.',
                ),
            )
            for field_name in field_names
        }
        wire_model = create_model(f"{fields.__name__}Wire", __base__=fields, **extra_fields)

    return DecisionSchema(
        name=name,
        version=version,
        confidence_strategy=confidence,
        wire_model=wire_model,
        values_model=fields,
        field_names=field_names,
    )

File: packages/core/src/schema.ts
import { z } from "zod";
import type { ConfidenceStrategy } from "./types.js";

export interface DecisionSchema<Shape extends z.ZodRawShape = z.ZodRawShape> {
  name: string;
  version: string;
  confidenceStrategy: ConfidenceStrategy;
  /**
   * Plain Zod object usable anywhere a ZodSchema is expected (generateObject,
   * withStructuredOutput, etc). When confidenceStrategy is "self-reported" this
   * carries extra `${field}_confidence` keys at runtime that callers outside
   * jevcal can safely ignore; the router strips them back out.
   */
  schema: z.ZodObject<Shape>;
  /** The original, un-augmented field shape — always safe to `.parse()` a backend's cleaned value object against. */
  valuesSchema: z.ZodObject<Shape>;
  fieldNames: (keyof Shape)[];
}

export interface DefineDecisionOptions<Shape extends z.ZodRawShape> {
  name: string;
  /** Schema version — segments calibration data and unlocks per-version drift tracking later. */
  version?: string;
  /** Defaults to "self-reported" since it's the only strategy portable to every backend. */
  confidence?: ConfidenceStrategy;
  fields: z.ZodObject<Shape>;
}

export function confidenceFieldKey(field: string | number | symbol): string {
  return `${String(field)}_confidence`;
}

export function defineDecision<Shape extends z.ZodRawShape>(
  options: DefineDecisionOptions<Shape>,
): DecisionSchema<Shape> {
  const confidenceStrategy = options.confidence ?? "self-reported";
  const fieldNames = Object.keys(options.fields.shape) as (keyof Shape)[];

  let schema: z.ZodObject<z.ZodRawShape> = options.fields;
  if (confidenceStrategy === "self-reported") {
    const confidenceFields: Record<string, z.ZodTypeAny> = {};
    for (const key of fieldNames) {
      confidenceFields[confidenceFieldKey(key)] = z
        .number()
        .min(0)
        .max(1)
        .describe(`Self-rated confidence (0-1) that the "${String(key)}" field above is correct.`);
    }
    schema = options.fields.extend(confidenceFields);
  }

  return {
    name: options.name,
    version: options.version ?? "1",
    confidenceStrategy,
    schema: schema as z.ZodObject<Shape>,
    valuesSchema: options.fields,
    fieldNames,
  };
}
